concent_api/core/payments/report.py

```diff
@@ -0,0 +1,759 @@
+    return [
+        forced_subtask_payment for forced_subtask_payment in forced_subtask_payments
+        if forced_subtask_payment.subtask_id == subtask_results_settled.subtask.subtask_id
+    ]
+
+
+def find_subtask_result_for_additional_verification_payment(
+    additional_verification_payment: CoverAdditionalVerificationEvent,
+    subtask_results_settled: list,
+    subtask_results_rejected: list,
+) -> List[PendingResponse]:
+    """ Find the SubtaskResultsSettled or SubtaskResultsRejected for given additional verification payment. """
+    matched_subtask_results_settled = [
+        results_settled for results_settled in subtask_results_settled
+        if results_settled.subtask.subtask_id == additional_verification_payment.subtask_id
+    ]
+
+    matched_subtask_results_rejected = [
+        results_rejected for results_rejected in subtask_results_rejected
+        if results_rejected.subtask.subtask_id == additional_verification_payment.subtask_id
+    ]
+
+    return matched_subtask_results_settled + matched_subtask_results_rejected
+
+
+def find_settlement_payment_for_force_payment_committed(
+    force_payment_committed_pending_response: PendingResponse,
+    settlement_payments: list,
+) -> List[ForcedPaymentEvent]:
+    """ Find settlement payment for given ForcePaymentCommitted message. """
+    payment_info = force_payment_committed_pending_response.payments.order_by('id').last()
+    return [
+        settlement_payment for settlement_payment in settlement_payments
+        if (
+            int(parse_datetime_to_timestamp(payment_info.payment_ts)) - 10 <=
+            settlement_payment.closure_time <=
+            int(parse_datetime_to_timestamp(payment_info.payment_ts)) + 10
+        )
+    ]
+
+
+def get_errors_and_warnings(payments: Payments, responses: PendingResponses) -> Tuple[List[Error], List[Warnings]]:
+    """ Get errors and warnings for messages and related payments. """
+    errors: List[Error] = []
+    warnings: List[Warnings] = []
+
+    for subtask_results_settled_pending_response in responses.subtask_results_settled:
+        matched_forced_subtask_payments_with_subtask_results_settled = find_forced_subtask_payments_for_subtask_results_settled(
+            subtask_results_settled_pending_response,
+            payments.forced_subtask_payments,
+        )
+
+        error_was_added = check_error_cannot_match_message_with_any_payment(
+            errors,
+            matched_forced_subtask_payments_with_subtask_results_settled,
+            subtask_results_settled_pending_response,
+        )
+
+        if error_was_added:
+            check_warning_deposit_claim_exists_for_missing_match(
+                warnings,
+                subtask_results_settled_pending_response.subtask.subtask_id,
+                ConcentUseCase.FORCED_ACCEPTANCE,
+            )
+
+        check_error_more_matching_payment_for_message_than_given_amount(
```

Comment:
I am not a fan of changing parameter inside function and not returning it, but maybe here you had nire unoirtabt goal to achieve :)

Reply:
Had what goal? "nire unoirtabt"

Reply:
I agree. I think the function should return the errors/warning objects.

Reply:
How would like to make the call then? All functions are adding the error conditionally.

```
error = check_x()
if error:
    errors.append(error)
```

This is much less readable.

```
errors.append(check_x())
```

This results in error looking like `[None]`.